### PR TITLE
Add support for specifying build args to use for docker build

### PR DESCRIPTION
**Pull request by hankjacobs**
Add support for specifying build args to use for docker build

### CODE REVIEW

**Comment by DavidHuie** [comment]
Can we pass in the Git SHA by default?

**Comment by hankjacobs** [comment]
We could. I think the question is "should we?". IMO, no. Keep everything as minimal as possible.

**Comment by DavidHuie** [comment]
The motivation for this PR is that exact feature, no? I'm glad that this solves it in the general case, but I think it would also be helpful to have a build argument like `FURAN_PROJECT_GIT_REF` always available.

File: cmd/build.go [hunk 1]
```diff
@@ -88,6 +91,21 @@ func validateCLIBuildRequest() {
 	}
 }
 
+func buildArgsFromSlice(args []string) map[string]string {
+	fmt.Println(args)
```
Comment (bkeroackdsc) [inline cmd/build.go]: Debugging printfs?

File: cmd/trigger.go [hunk 2]
```diff
@@ -48,6 +49,7 @@ func init() {
 	triggerCmd.PersistentFlags().StringVar(&tags, "tags", "master", "image tags (optional, comma-delimited)")
 	triggerCmd.PersistentFlags().BoolVar(&cliBuildRequest.Build.TagWithCommitSha, "tag-sha", false, "additionally tag with git commit SHA (optional)")
 	triggerCmd.PersistentFlags().BoolVar(&cliBuildRequest.SkipIfExists, "skip-if-exists", false, "if build already exists at destination, skip build/push (registry: all tags exist, s3: object exists)")
+	triggerCmd.PersistentFlags().StringSliceVar(&buildArgs, "build-arg", nil, "Build arg to use for build request")
```
Comment (bkeroackdsc) [inline cmd/trigger.go]: Perhaps empty slice instead of nil for the default.

**Comment by bkeroackdsc** [comment]
I agree with Hank. We don't want any magic in Furan. The build performed should be exactly reproducible locally with standard Docker tooling.

File: cmd/build.go (hunk 1 above)
Comment (hankjacobs) [reply cmd/build.go]: blah I thought I took that out. Good catch. ty.

File: cmd/trigger.go (hunk 2 above)
Comment (hankjacobs) [reply cmd/trigger.go]: :+1:

**Comment by DavidHuie** [comment]
Including a build argument doesn't affect the build process. The Dockerfile has to opt-in to using the build argument.

**Comment by hankjacobs** [comment]
It pollutes the build context with an arg that only might be used

**Comment by hankjacobs** [comment]
I really prefer it to be an opt-in thing. It was the inspiration for this PR but this PR is very much a generic implementation (arbitrary build args).

**Comment by DavidHuie** [comment]
True, but is polluting the build context something that we have to worry about? I don't think there are any issues with introducing an optionally used build argument.

Furan builds Docker images fetched from Github, specified by a Git ref. Thus, it's definitely more than possible that this feature could be useful to many since it's one of the parameters specified when requesting the build. This is similar to the Kubernetes downward API, which makes pod metadata optionally available to pods.

**Comment by bkeroackdsc** [comment]
If the Dockerfile assumes magic args are set it will break (or require extra steps) when running the build locally.

**Comment by DavidHuie** [comment]
@bkeroackdsc I don't understand the concern. There aren't any consequences to having an unused build argument. If the user specifies a build argument, they also specify a default. So if someone opted in to the Furan build argument, they would also have a default for local builds.

**Comment by DavidHuie** [comment]
Anyway, my comment shouldn't block the merging of this PR.

**Comment by hankjacobs** [comment]
@bkeroackdsc Feedback implemented

**Comment by DavidHuie** [comment]
Just found some precedent to predefined build args: https://docs.docker.com/engine/reference/builder/#predefined-args

Docker includes several be default.

**Comment by hankjacobs** [comment]
I move to punt on that and merge the PR as is. We can revisit it outside of this PR. Cool with you @bkeroackdsc @DavidHuie?

**Comment by bkeroackdsc** [comment]
@DavidHuie I get what you're saying with the downward API analogy, and I think that idea is interesting. Let's merge this PR as-is right now and come up with a "downward args" spec that can be implemented.

**Pull request merged**